# Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„ØµØ§Ø±Ù…Ø© Ù„ØªØ±Ø§Ø¨Ø· Ù…Ø³ØªÙˆØ¯Ø¹Ø§Øª SEMOP

**Ø§Ù„Ø¥ØµØ¯Ø§Ø±:** 1.0  
**ØªØ§Ø±ÙŠØ® Ø§Ù„Ø¥Ù†Ø´Ø§Ø¡:** 3 Ø¯ÙŠØ³Ù…Ø¨Ø± 2025  
**Ø§Ù„Ø­Ø§Ù„Ø©:** Ø¥Ù„Ø²Ø§Ù…ÙŠ - Ù„Ø§ Ø§Ø³ØªØ«Ù†Ø§Ø¡Ø§Øª

---

## âš–ï¸ Ø§Ù„Ù…Ø¨Ø¯Ø£ Ø§Ù„Ø£Ø³Ø§Ø³ÙŠ

**"Ø§Ù„ØªØ±Ø§Ø¨Ø· Ø§Ù„Ø­Ù‚ÙŠÙ‚ÙŠ ØºÙŠØ± Ù‚Ø§Ø¨Ù„ Ù„Ù„ØªÙØ§ÙˆØ¶"**

Ù‡Ø°Ù‡ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ù…ØµÙ…Ù…Ø© Ù„Ø¶Ù…Ø§Ù† Ø£Ù† Ø§Ù„Ù…Ø³ØªÙˆØ¯Ø¹Ø§Øª Ø§Ù„Ø£Ø±Ø¨Ø¹Ø© ØªØ¹Ù…Ù„ ÙƒÙˆØ­Ø¯Ø© ÙˆØ§Ø­Ø¯Ø© Ù…ØªÙ…Ø§Ø³ÙƒØ©ØŒ ÙˆÙ„Ù…Ù†Ø¹ Ø£ÙŠ ØªÙÙƒÙƒ Ø£Ùˆ ØªØ¹Ø§Ø±Ø¶ Ù…Ø³ØªÙ‚Ø¨Ù„Ø§Ù‹.

---

## ğŸ”´ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ø­Ø±Ø¬Ø© (Critical Rules)

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 1: Ø§Ù„Ø¹Ù‚ÙˆØ¯ Ø§Ù„Ù…Ø´ØªØ±ÙƒØ© Ù‡ÙŠ Ø§Ù„Ù‚Ø§Ù†ÙˆÙ†**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… **ÙŠØ¬Ø¨** Ø§Ø³ØªØ®Ø¯Ø§Ù… `@semop/contracts` Ù„Ø¬Ù…ÙŠØ¹ DTOs, Enums, Interfaces
- âŒ **ÙŠÙ…Ù†Ø¹** Ø¥Ù†Ø´Ø§Ø¡ DTOs Ù…Ø­Ù„ÙŠØ© ÙÙŠ Ø£ÙŠ Ù…Ø³ØªÙˆØ¯Ø¹
- âŒ **ÙŠÙ…Ù†Ø¹** Ù†Ø³Ø® Ø§Ù„Ø£Ù†ÙˆØ§Ø¹ Ù…Ù† Ø§Ù„Ø¹Ù‚ÙˆØ¯ Ø§Ù„Ù…Ø´ØªØ±ÙƒØ©

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```typescript
// âœ… ØµØ­ÙŠØ­
import { UserDto, CreateUserDto } from '@semop/contracts';

// âŒ Ø®Ø·Ø£ - Ù…Ù…Ù†ÙˆØ¹
interface UserDto {
  id: string;
  email: string;
}
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
# ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ…Ø± Ù‡Ø°Ø§ Ø§Ù„ÙØ­Øµ Ù‚Ø¨Ù„ ÙƒÙ„ commit
./scripts/verify-no-local-dtos.sh
```

#### **Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø©:**
- Ø±ÙØ¶ Pull Request ØªÙ„Ù‚Ø§Ø¦ÙŠØ§Ù‹
- Ø¥Ø¹Ø§Ø¯Ø© ÙƒØªØ§Ø¨Ø© Ø§Ù„ÙƒÙˆØ¯

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 2: Ù†ÙØ³ Ø§Ù„Ø¥ØµØ¯Ø§Ø± Ø¯Ø§Ø¦Ù…Ø§Ù‹**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ø³ØªÙˆØ¯Ø¹Ø§Øª ØªØ³ØªØ®Ø¯Ù… Ù†ÙØ³ Ø¥ØµØ¯Ø§Ø± `@semop/contracts`
- âœ… Ø§Ù„ØªØ­Ø¯ÙŠØ« ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙƒÙˆÙ† Ù…ØªØ²Ø§Ù…Ù†
- âŒ **ÙŠÙ…Ù†Ø¹** Ø§Ø³ØªØ®Ø¯Ø§Ù… Ø¥ØµØ¯Ø§Ø±Ø§Øª Ù…Ø®ØªÙ„ÙØ©

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```json
// unified-backend-monorepo/package.json
{
  "dependencies": {
    "@semop/contracts": "1.2.3"  // Ù†ÙØ³ Ø§Ù„Ø¥ØµØ¯Ø§Ø± Ø¨Ø§Ù„Ø¶Ø¨Ø·
  }
}

// unified-frontend-monorepo/package.json
{
  "dependencies": {
    "@semop/contracts": "1.2.3"  // Ù†ÙØ³ Ø§Ù„Ø¥ØµØ¯Ø§Ø± Ø¨Ø§Ù„Ø¶Ø¨Ø·
  }
}
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
./scripts/check-contracts-version.sh
# ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ¹Ø±Ø¶: âœ… All repos use version 1.2.3
```

#### **Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø©:**
- Ù…Ù†Ø¹ Ø§Ù„Ù†Ø´Ø±
- ØªÙ†Ø¨ÙŠÙ‡ ÙÙˆØ±ÙŠ

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 3: Ø§Ù„Ø¨Ù†ÙŠØ© Ø§Ù„Ø·Ø¨Ù‚ÙŠØ© Ø¥Ù„Ø²Ø§Ù…ÙŠØ©**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ù†Ø·Ù‚ Ø§Ù„ØªØ¬Ø§Ø±ÙŠ ÙÙŠ `libs`
- âœ… `apps` Ù„Ù„ØªØ·Ø¨ÙŠÙ‚Ø§Øª ÙÙ‚Ø· (thin layer)
- âŒ **ÙŠÙ…Ù†Ø¹** ÙˆØ¶Ø¹ Ù…Ù†Ø·Ù‚ ØªØ¬Ø§Ø±ÙŠ ÙÙŠ `apps`

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```
unified-backend-monorepo/
â”œâ”€â”€ apps/
â”‚   â””â”€â”€ api-gateway/          # ÙÙ‚Ø· routing Ùˆ configuration
â”‚       â””â”€â”€ src/
â”‚           â””â”€â”€ main.ts       # bootstrap ÙÙ‚Ø·
â””â”€â”€ libs/
    â”œâ”€â”€ 1-core-services/      # Ø§Ù„Ù…Ù†Ø·Ù‚ Ø§Ù„ØªØ¬Ø§Ø±ÙŠ Ù‡Ù†Ø§
    â”œâ”€â”€ 2-operational-platform/
    â”œâ”€â”€ 3-vertical-applications/
    â””â”€â”€ 4-sector-libraries/
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
# ÙŠØ¬Ø¨ Ø£Ù„Ø§ ÙŠÙˆØ¬Ø¯ Ù…Ù„ÙØ§Øª service ÙÙŠ apps
find apps -name "*service.ts" | wc -l
# Ø§Ù„Ù†ØªÙŠØ¬Ø© Ø§Ù„Ù…ØªÙˆÙ‚Ø¹Ø©: 0
```

#### **Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø©:**
- Ø±ÙØ¶ Pull Request
- Ø¥Ø¹Ø§Ø¯Ø© Ù‡ÙŠÙƒÙ„Ø© Ø¥Ù„Ø²Ø§Ù…ÙŠØ©

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 4: ÙƒÙˆÙ…Ù†Øª Ù„ÙƒÙ„ ØªØºÙŠÙŠØ±**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… ÙƒÙ„ Ù…Ù„Ù Ù…Ø¹Ø¯Ù„ ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ ÙƒÙˆÙ…Ù†Øª Ø¨Ø§Ù„Ù…Ø±Ø­Ù„Ø©
- âœ… Ø§Ù„ÙƒÙˆÙ…Ù†Øª ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙƒÙˆÙ† ÙÙŠ Ø£Ø¹Ù„Ù‰ Ø§Ù„Ù…Ù„Ù Ø£Ùˆ Ù‚Ø¨Ù„ Ø§Ù„ØªØºÙŠÙŠØ±
- âŒ **ÙŠÙ…Ù†Ø¹** Ø§Ù„ØªØºÙŠÙŠØ± Ø¨Ø¯ÙˆÙ† ØªÙˆØ«ÙŠÙ‚

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```typescript
/**
 * PHASE-2.3.1: Migrated to use @semop/contracts
 * Previous: Local UserDto interface
 * Current: Imported from @semop/contracts
 * Date: 2025-12-10
 * Author: Development Team
 */
import { UserDto } from '@semop/contracts';

export class UsersService {
  // PHASE-2.3.1: Updated method signature to use shared contract
  async findAll(): Promise<UserDto[]> {
    // ...
  }
}
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
# Ø§Ù„ØªØ­Ù‚Ù‚ Ù…Ù† ÙˆØ¬ÙˆØ¯ PHASE ÙÙŠ Ø§Ù„Ù…Ù„ÙØ§Øª Ø§Ù„Ù…Ø¹Ø¯Ù„Ø©
git diff --name-only | xargs grep -L "PHASE-" && echo "âŒ Missing PHASE comment"
```

#### **Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø©:**
- Ø±ÙØ¶ Commit
- Ø·Ù„Ø¨ Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„ØªÙˆØ«ÙŠÙ‚

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 5: Ù„Ø§ merge Ø¨Ø¯ÙˆÙ† ÙØ­Øµ**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… Ø¬Ù…ÙŠØ¹ Ø§Ù„Ø§Ø®ØªØ¨Ø§Ø±Ø§Øª ÙŠØ¬Ø¨ Ø£Ù† ØªÙ…Ø±
- âœ… Linting ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
- âœ… Build ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
- âœ… Ù…Ø±Ø§Ø¬Ø¹Ø© ÙƒÙˆØ¯ Ù…Ù† Ø´Ø®Øµ Ø¢Ø®Ø±
- âŒ **ÙŠÙ…Ù†Ø¹** Merge Ù…Ø¨Ø§Ø´Ø± Ø¥Ù„Ù‰ main

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```yaml
# .github/workflows/pr-checks.yml
name: PR Checks
on: pull_request
jobs:
  checks:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - run: npm ci
      - run: npm run test          # âœ… ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
      - run: npm run lint          # âœ… ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
      - run: npm run build         # âœ… ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
      - run: ./scripts/verify-integration.sh  # âœ… ÙŠØ¬Ø¨ Ø£Ù† ÙŠÙ†Ø¬Ø­
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
- GitHub Actions ØªÙ„Ù‚Ø§Ø¦ÙŠØ§Ù‹
- Branch Protection Rules

#### **Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø©:**
- Ù…Ù†Ø¹ Merge ØªÙ„Ù‚Ø§Ø¦ÙŠØ§Ù‹
- Ø·Ù„Ø¨ Ø¥ØµÙ„Ø§Ø­ Ø§Ù„Ø£Ø®Ø·Ø§Ø¡

---

## ğŸŸ¡ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…Ù‡Ù…Ø© (Important Rules)

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 6: Ø§Ù„ØªØ³Ù…ÙŠØ© Ø§Ù„Ù…ÙˆØ­Ø¯Ø©**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… Ø§Ø³ØªØ®Ø¯Ø§Ù… camelCase Ù„Ù„Ù…ØªØºÙŠØ±Ø§Øª ÙˆØ§Ù„Ø¯ÙˆØ§Ù„
- âœ… Ø§Ø³ØªØ®Ø¯Ø§Ù… PascalCase Ù„Ù„ÙƒÙ„Ø§Ø³Ø§Øª ÙˆØ§Ù„ÙˆØ§Ø¬Ù‡Ø§Øª
- âœ… Ø§Ø³ØªØ®Ø¯Ø§Ù… kebab-case Ù„Ø£Ø³Ù…Ø§Ø¡ Ø§Ù„Ù…Ù„ÙØ§Øª
- âœ… Ø§Ø³ØªØ®Ø¯Ø§Ù… UPPER_SNAKE_CASE Ù„Ù„Ø«ÙˆØ§Ø¨Øª

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```typescript
// âœ… ØµØ­ÙŠØ­
const userName = 'John';
const MAX_RETRY_COUNT = 3;
class UserService {}
interface UserDto {}

// âŒ Ø®Ø·Ø£
const UserName = 'John';
const maxRetryCount = 3;
class userService {}
interface user_dto {}
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
npm run lint
```

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 7: Ø§Ù„ØªÙˆØ«ÙŠÙ‚ Ø§Ù„Ø¥Ù„Ø²Ø§Ù…ÙŠ**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… ÙƒÙ„ Ø¯Ø§Ù„Ø© Ø¹Ø§Ù…Ø© ÙŠØ¬Ø¨ Ø£Ù† ØªØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ JSDoc
- âœ… ÙƒÙ„ ÙƒÙ„Ø§Ø³ ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ ÙˆØµÙ
- âœ… ÙƒÙ„ API endpoint ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ Swagger decorator

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```typescript
/**
 * Creates a new user in the system
 * @param createUserDto - User creation data
 * @returns Created user with generated ID
 * @throws BadRequestException if email already exists
 */
@ApiOperation({ summary: 'Create new user' })
@ApiResponse({ status: 201, description: 'User created successfully', type: UserDto })
@ApiResponse({ status: 400, description: 'Invalid input' })
async createUser(@Body() createUserDto: CreateUserDto): Promise<UserDto> {
  // ...
}
```

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 8: Ø§Ù„Ø§Ø®ØªØ¨Ø§Ø±Ø§Øª Ø¥Ù„Ø²Ø§Ù…ÙŠØ©**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… ÙƒÙ„ service ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ unit tests
- âœ… ÙƒÙ„ controller ÙŠØ¬Ø¨ Ø£Ù† ÙŠØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ integration tests
- âœ… Coverage Ù„Ø§ ÙŠÙ‚Ù„ Ø¹Ù† 80%

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```typescript
// users.service.spec.ts
describe('UsersService', () => {
  let service: UsersService;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [UsersService],
    }).compile();
    service = module.get<UsersService>(UsersService);
  });

  it('should create user', async () => {
    const dto: CreateUserDto = { email: 'test@example.com' };
    const result = await service.create(dto);
    expect(result).toBeDefined();
    expect(result.email).toBe(dto.email);
  });
});
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```bash
npm run test:cov
# Coverage threshold: 80%
```

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 9: Git Workflow Ø§Ù„ØµØ§Ø±Ù…**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… ÙØ±Ø¹ `main` Ù…Ø­Ù…ÙŠ - Ù„Ø§ commits Ù…Ø¨Ø§Ø´Ø±Ø©
- âœ… Ø¬Ù…ÙŠØ¹ Ø§Ù„ØªØºÙŠÙŠØ±Ø§Øª Ø¹Ø¨Ø± Pull Requests
- âœ… ØªØ³Ù…ÙŠØ© Ø§Ù„ÙØ±ÙˆØ¹: `feature/`, `fix/`, `refactor/`
- âœ… Commit messages ÙˆØ§Ø¶Ø­Ø© ÙˆÙ…ÙØµÙ„Ø©

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```bash
# âœ… ØµØ­ÙŠØ­
git checkout -b feature/add-billing-service
git commit -m "feat(billing): add invoice generation service

- Implemented InvoiceService with PDF generation
- Added unit tests with 95% coverage
- Updated API documentation
- PHASE-2.2.1"

# âŒ Ø®Ø·Ø£
git checkout -b my-changes
git commit -m "updates"
```

#### **Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ:**
```yaml
# .github/workflows/commit-lint.yml
- uses: wagoid/commitlint-github-action@v5
```

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 10: Ø§Ù„Ù…Ø²Ø§Ù…Ù†Ø© Ø§Ù„ÙŠÙˆÙ…ÙŠØ©**

#### **Ø§Ù„Ø¥Ù„Ø²Ø§Ù…:**
- âœ… Push ÙŠÙˆÙ…ÙŠ Ø¹Ù„Ù‰ Ø§Ù„Ø£Ù‚Ù„
- âœ… ØªØ­Ø¯ÙŠØ« `PROGRESS_TRACKER.md` ÙŠÙˆÙ…ÙŠØ§Ù‹
- âœ… Ù…Ø²Ø§Ù…Ù†Ø© Ù…Ø¹ main Ù‚Ø¨Ù„ Ø§Ù„Ø¨Ø¯Ø¡ Ø¨Ø§Ù„Ø¹Ù…Ù„

#### **Ø§Ù„ØªØ·Ø¨ÙŠÙ‚:**
```bash
# ÙƒÙ„ ØµØ¨Ø§Ø­
git checkout main
git pull origin main
git checkout feature/my-feature
git rebase main

# ÙƒÙ„ Ù…Ø³Ø§Ø¡
git add .
git commit -m "chore: daily progress update"
git push origin feature/my-feature

# ØªØ­Ø¯ÙŠØ« PROGRESS_TRACKER.md
echo "## $(date +%Y-%m-%d)" >> PROGRESS_TRACKER.md
echo "- âœ… Completed task X" >> PROGRESS_TRACKER.md
```

---

## ğŸŸ¢ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…ÙˆØµÙ‰ Ø¨Ù‡Ø§ (Recommended Rules)

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 11: Ø§Ù„Ø£Ø¯Ø§Ø¡ ÙˆØ§Ù„ØªØ­Ø³ÙŠÙ†**

- Ø§Ø³ØªØ®Ø¯Ø§Ù… Caching Ø­ÙŠØ«Ù…Ø§ Ø£Ù…ÙƒÙ†
- ØªØ­Ø³ÙŠÙ† Ø§Ù„Ø§Ø³ØªØ¹Ù„Ø§Ù…Ø§Øª (Query Optimization)
- Lazy Loading Ù„Ù„Ø¨ÙŠØ§Ù†Ø§Øª Ø§Ù„ÙƒØ¨ÙŠØ±Ø©
- Pagination Ù„Ù„Ù‚ÙˆØ§Ø¦Ù…

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 12: Ø§Ù„Ø£Ù…Ø§Ù†**

- Ø¹Ø¯Ù… ØªØ®Ø²ÙŠÙ† Secrets ÙÙŠ Ø§Ù„ÙƒÙˆØ¯
- Ø§Ø³ØªØ®Ø¯Ø§Ù… Environment Variables
- Validation Ù„Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ø¯Ø®Ù„Ø§Øª
- Authentication & Authorization Ù„Ø¬Ù…ÙŠØ¹ APIs

---

### **Ø§Ù„Ù‚Ø§Ø¹Ø¯Ø© 13: Ù‚Ø§Ø¨Ù„ÙŠØ© Ø§Ù„ØµÙŠØ§Ù†Ø©**

- DRY (Don't Repeat Yourself)
- SOLID Principles
- Clean Code
- Refactoring Ù…Ø³ØªÙ…Ø±

---

## ğŸ”§ Ø£Ø¯ÙˆØ§Øª Ø§Ù„ÙØ­Øµ Ø§Ù„Ø¢Ù„ÙŠ

### **1. Pre-commit Hooks**

```bash
# .husky/pre-commit
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

# Check for PHASE comments
echo "ğŸ” Checking for PHASE comments..."
./scripts/check-phase-comments.sh || exit 1

# Run linting
echo "ğŸ” Running linter..."
npm run lint || exit 1

# Run tests
echo "ğŸ§ª Running tests..."
npm run test || exit 1

echo "âœ… All checks passed!"
```

### **2. Pre-push Hooks**

```bash
# .husky/pre-push
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

# Check contracts version
echo "ğŸ” Checking contracts version..."
./scripts/check-contracts-version.sh || exit 1

# Verify integration
echo "ğŸ” Verifying integration..."
./scripts/verify-integration.sh || exit 1

echo "âœ… Ready to push!"
```

### **3. GitHub Actions**

```yaml
# .github/workflows/strict-checks.yml
name: Strict Checks
on: [push, pull_request]
jobs:
  enforce-rules:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      
      - name: Check contracts version
        run: ./scripts/check-contracts-version.sh
      
      - name: Check for local DTOs
        run: ./scripts/verify-no-local-dtos.sh
      
      - name: Check layer architecture
        run: ./scripts/verify-layer-architecture.sh
      
      - name: Check PHASE comments
        run: ./scripts/check-phase-comments.sh
      
      - name: Run tests
        run: npm run test:cov
      
      - name: Check coverage
        run: |
          COVERAGE=$(cat coverage/coverage-summary.json | jq '.total.lines.pct')
          if (( $(echo "$COVERAGE < 80" | bc -l) )); then
            echo "âŒ Coverage is below 80%: $COVERAGE%"
            exit 1
          fi
```

---

## ğŸ“Š Ù…Ø¤Ø´Ø±Ø§Øª Ø§Ù„Ø§Ù„ØªØ²Ø§Ù…

### **ÙŠÙˆÙ…ÙŠ:**
- [ ] Ø¬Ù…ÙŠØ¹ Commits ØªØ­ØªÙˆÙŠ Ø¹Ù„Ù‰ PHASE comments
- [ ] Ø¬Ù…ÙŠØ¹ Tests ØªÙ…Ø±
- [ ] Linting ÙŠÙ†Ø¬Ø­
- [ ] PROGRESS_TRACKER Ù…Ø­Ø¯Ø«

### **Ø£Ø³Ø¨ÙˆØ¹ÙŠ:**
- [ ] Coverage Ø£Ø¹Ù„Ù‰ Ù…Ù† 80%
- [ ] Ù„Ø§ ØªÙˆØ¬Ø¯ DTOs Ù…Ø­Ù„ÙŠØ©
- [ ] Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ø³ØªÙˆØ¯Ø¹Ø§Øª Ø¹Ù„Ù‰ Ù†ÙØ³ Ø¥ØµØ¯Ø§Ø± Ø§Ù„Ø¹Ù‚ÙˆØ¯
- [ ] Ø§Ù„Ø¨Ù†ÙŠØ© Ø§Ù„Ø·Ø¨Ù‚ÙŠØ© Ù…Ø·Ø¨Ù‚Ø© 100%

### **Ø´Ù‡Ø±ÙŠ:**
- [ ] Documentation ÙƒØ§Ù…Ù„
- [ ] Performance Tests ØªÙ…Ø±
- [ ] Security Audit Ù†Ø¸ÙŠÙ
- [ ] Refactoring Debt Ù…Ù†Ø®ÙØ¶

---

## âš ï¸ Ø§Ù„Ø¹Ù‚ÙˆØ¨Ø§Øª

### **Ù…Ø®Ø§Ù„ÙØ© Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ø­Ø±Ø¬Ø©:**
1. Ø±ÙØ¶ Pull Request ØªÙ„Ù‚Ø§Ø¦ÙŠØ§Ù‹
2. Ù…Ù†Ø¹ Merge
3. Ø·Ù„Ø¨ Ø¥Ø¹Ø§Ø¯Ø© ÙƒØªØ§Ø¨Ø© Ø§Ù„ÙƒÙˆØ¯
4. ØªÙ†Ø¨ÙŠÙ‡ Ø§Ù„ÙØ±ÙŠÙ‚

### **Ù…Ø®Ø§Ù„ÙØ© Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…Ù‡Ù…Ø©:**
1. ØªØ­Ø°ÙŠØ± ÙÙŠ PR
2. Ø·Ù„Ø¨ Ø¥ØµÙ„Ø§Ø­
3. Ù…Ø±Ø§Ø¬Ø¹Ø© Ø¥Ø¶Ø§ÙÙŠØ©

### **Ù…Ø®Ø§Ù„ÙØ© Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…ÙˆØµÙ‰ Ø¨Ù‡Ø§:**
1. Ù…Ù„Ø§Ø­Ø¸Ø© ÙÙŠ PR
2. Ø§Ù‚ØªØ±Ø§Ø­ ØªØ­Ø³ÙŠÙ†

---

## ğŸ¯ Ø§Ù„Ø§Ù„ØªØ²Ø§Ù…

**Ø¨ØªÙˆÙ‚ÙŠØ¹ÙŠ Ø¹Ù„Ù‰ Ù‡Ø°Ø§ Ø§Ù„Ù…Ø³ØªÙ†Ø¯ØŒ Ø£ØªØ¹Ù‡Ø¯ Ø¨Ù€:**

1. âœ… Ø§Ù„Ø§Ù„ØªØ²Ø§Ù… Ø¨Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ø­Ø±Ø¬Ø© Ø¨Ø¯ÙˆÙ† Ø§Ø³ØªØ«Ù†Ø§Ø¡
2. âœ… Ø¨Ø°Ù„ Ø£Ù‚ØµÙ‰ Ø¬Ù‡Ø¯ Ù„Ù„Ø§Ù„ØªØ²Ø§Ù… Ø¨Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…Ù‡Ù…Ø©
3. âœ… Ù…Ø±Ø§Ø¹Ø§Ø© Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ø§Ù„Ù…ÙˆØµÙ‰ Ø¨Ù‡Ø§ Ù‚Ø¯Ø± Ø§Ù„Ø¥Ù…ÙƒØ§Ù†
4. âœ… Ø§Ù„Ù…Ø³Ø§Ù‡Ù…Ø© ÙÙŠ ØªØ­Ø³ÙŠÙ† Ù‡Ø°Ù‡ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯
5. âœ… Ø§Ù„Ø¥Ø¨Ù„Ø§Øº Ø¹Ù† Ø£ÙŠ Ù…Ø®Ø§Ù„ÙØ§Øª

**Ø§Ù„ØªÙˆÙ‚ÙŠØ¹ Ø§Ù„Ø±Ù‚Ù…ÙŠ:**
```
Repository: SEMOP
Date: 2025-12-03
Commit: [Ø³ÙŠØªÙ… Ø¥Ø¶Ø§ÙØªÙ‡ Ø¹Ù†Ø¯ Ø§Ù„ØªØ·Ø¨ÙŠÙ‚]
```

---

## ğŸ“ Ø§Ù„Ø¥Ø¨Ù„Ø§Øº Ø¹Ù† Ø§Ù„Ù…Ø´Ø§ÙƒÙ„

Ø¥Ø°Ø§ ÙˆØ§Ø¬Ù‡Øª Ø£ÙŠ ØµØ¹ÙˆØ¨Ø© ÙÙŠ ØªØ·Ø¨ÙŠÙ‚ Ù‡Ø°Ù‡ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯:

1. Ø§ÙØªØ­ Issue ÙÙŠ Ù…Ø³ØªÙˆØ¯Ø¹ SEMOP
2. Ø§Ø³ØªØ®Ø¯Ù… Label: `strict-rules`
3. Ø§Ø´Ø±Ø­ Ø§Ù„Ù…Ø´ÙƒÙ„Ø© Ø¨Ø§Ù„ØªÙØµÙŠÙ„
4. Ø§Ù‚ØªØ±Ø­ Ø­Ù„Ø§Ù‹ Ø¥Ù† Ø£Ù…ÙƒÙ†

---

## ğŸ”„ ØªØ­Ø¯ÙŠØ« Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯

Ù‡Ø°Ù‡ Ø§Ù„Ù‚ÙˆØ§Ø¹Ø¯ Ù‚Ø§Ø¨Ù„Ø© Ù„Ù„ØªØ­Ø¯ÙŠØ« Ø¨Ù†Ø§Ø¡Ù‹ Ø¹Ù„Ù‰:
- Ø§Ù„ØªØ¬Ø±Ø¨Ø© Ø§Ù„Ø¹Ù…Ù„ÙŠØ©
- Ø§Ù„ØªØ­Ø¯ÙŠØ§Øª Ø§Ù„Ù…ÙˆØ§Ø¬Ù‡Ø©
- Ø£ÙØ¶Ù„ Ø§Ù„Ù…Ù…Ø§Ø±Ø³Ø§Øª Ø§Ù„Ø¬Ø¯ÙŠØ¯Ø©

**Ø¹Ù…Ù„ÙŠØ© Ø§Ù„ØªØ­Ø¯ÙŠØ«:**
1. Ø§Ù‚ØªØ±Ø§Ø­ Ø§Ù„ØªØ¹Ø¯ÙŠÙ„ ÙÙŠ Issue
2. Ù…Ù†Ø§Ù‚Ø´Ø© Ù…Ø¹ Ø§Ù„ÙØ±ÙŠÙ‚
3. Ù…ÙˆØ§ÙÙ‚Ø© Ø¬Ù…Ø§Ø¹ÙŠØ©
4. ØªØ­Ø¯ÙŠØ« Ø§Ù„Ù…Ø³ØªÙ†Ø¯
5. Ø¥Ø´Ø¹Ø§Ø± Ø¬Ù…ÙŠØ¹ Ø§Ù„Ù…Ø·ÙˆØ±ÙŠÙ†

---

**Ø¢Ø®Ø± ØªØ­Ø¯ÙŠØ«:** 3 Ø¯ÙŠØ³Ù…Ø¨Ø± 2025  
**Ø§Ù„Ø¥ØµØ¯Ø§Ø±:** 1.0  
**Ø§Ù„Ø­Ø§Ù„Ø©:** Ù†Ø´Ø· ÙˆØ¥Ù„Ø²Ø§Ù…ÙŠ
